### ISSUE TITLE
custom scale for useSlider

### ISSUE THREAD

**Issue by agriffis** [adobe/react-spectrum#1830]
# 🙋 Feature Request

For sliders with a wide range of values, a logarithmic or other custom scale can be helpful to enable selecting higher precision for smaller values and lower precision for larger values.

## 🤔 Expected Behavior

The slider should feel linear, but output values based on a scale function. For example, the bottom of the slider might be zero, the middle might be 100, and the top might be 10,000.

## 😯 Current Behavior

Currently `useSlider` only does linear scale. 

## 💁 Possible Solution

It's possible to wrap `useSlider` with in/out transforms, but the `<output>` element in the slider is wrong, as will `aria-valuetext`, etc. It's better if `useSlider` is aware of the transform.

Material UI offers a `scale` prop: https://github.com/mui-org/material-ui/pull/19158/files

## 🔦 Context

Small biz might have 12 of something, where enterprise biz has 100,000 of the same thing. The small biz needs fine granularity at one end of the slider, and the enterprise biz needs large values available—even if they're not as precise—at the other end of the slider.

## 🧢 Your Company/Team

[Scampersand](https://scampersand.com) (not affiliated with Adobe)

**Comment by agriffis** [comment]
For anybody looking here for an interim solution, this is what I'm doing:

```js
import {useSliderState as useSliderStateRS} from 'react-stately'

// Wrap react-stately's useSliderState to apply scaling functions.
const useSliderState = ({toPercent, toValue, ...props}) => {
  let state = useSliderStateRS(props)

  if (toPercent && toValue) {
    return {
      ...state,
      getPercentValue: toValue,
      getValuePercent: toPercent,
      setThumbPercent: (index, percent) =>
        state.setThumbValue(index, toValue(percent)),
      getThumbPercent: index => toPercent(state.getThumbValue(index)),
    }
  }

  return state
}

// Make Slider with logarithmic scaling
<Slider
  toPercent={value => Math.log10(value + 1) / Math.log10(maxValue + 1)}
  toValue={percent => 10 ** (percent * Math.log10(maxValue + 1)) - 1}
/>
```

**Comment by snowystinger** [comment]
Hey! thanks for the issue. Just real quick, I think you're trying to do this with your own component, not the React Spectrum Slider, your own slider using the hooks? If so, you could just substitute whatever you want in place of what is returned by the hooks. Something like this:
https://codesandbox.io/s/little-monad-cx9l7?file=/src/App.js

If you're trying to do this with our Slider, then we'll need to check with our Design to make sure this is something we want to support. I think we'd want to explore some different API choices

**Comment by agriffis** [comment]
Hey @snowystinger, thanks for the reply.

You're right, I'm using `react-aria` and `react-stately` hooks with my own components. React Spectrum looks nice, but I'm particularly happy with the way you've architected the layers so people can bring their own presentation layer.

And yep, I can substitute what's returned by the hooks—indeed, that's what I'm doing in the code snippet above. Again, I'm really grateful for the architecture that makes this possible.

The downside is that it feels more tightly coupled to the `useSliderState` internals than I'd like. I had to walk through all the state properties to figure out what to substitute, and I'd be concerned that a minor update release of `react-stately` could break it.

**Comment by snowystinger** [comment]
Right, I see yours is coupled a bit. I think my suggestion is slightly less coupled, as it doesn't alter state being passed from stately to aria, it only deals with things in the render layer. I do have to know about aria properties, but those won't change unless html itself changes. Does that make more sense?

Also, I forgot to say that I'll bring this up with design next week.

**Comment by agriffis** [comment]
Sorry, I didn't mention that I looked at your codesandbox. Thanks for taking a hack at it!

Let's call your suggestion "value scaling" and mine "percent conversion."

The drawback to **value scaling** is that `useSliderState` tracks an internal value that's different from the external value, so you have to make sure to catch it and modify at every point. Anything that relies on the value, such as `defaultValue`, `minValue`, `maxValue`, `value` (controlled) and `step` are thrown off by the fact that the internal state is different from the external state. That's a lot of implementation details putting a burden on the caller.

By contrast, **percent conversion** modifies how `useSliderState` converts values to slider percentages and back. That means that internal and external values stay consistent, and you don't have to take any special measures as the caller.

Intuitively, it seems like there's a drawback to percent conversion, because you have to pass two functions, one for scaling (for determining the value as the slider moves) and the other for unscaling (for positioning the slider given a value). However value scaling actually has the same drawback, because the caller has to do those conversions for the props mentioned above:

<table><tr><th>value scaling</th><th>percent conversion</th></tr><td>

```js
// adjust by 1 to avoid -Infinity
const scale = intValue => 2 ** intValue - 1
const unscale = extValue => Math.log2(extValue + 1)

// logarithmic Slider [0, 100] defaulting to 50
<Slider
  scale={scale}
  minValue={unscale(0)}
  maxValue={unscale(100)}
  defaultvalue={unscale(50)}
/>
```

</td><td>

```js
// adjust by 1 to avoid -Infinity
const scale = intValue => 2 ** intValue - 1
const unscale = extValue => Math.log2(extValue + 1)

// logarithmic Slider [0, 100] defaulting to 50
<Slider
  scale={scale}
  unscale={unscale}
  minValue={0}
  maxValue={100}
  defaultvalue={50}
/>

// Slider converts scale/unscale → toValue/toPercent
// internally, so caller can pass simpler functions
```

</td></tr></table>

I'm thrilled to hear you'll bring this up with Design, thanks!

**Comment by snowystinger** [comment]
Ah, yes, I do see the issue for min/max etc, I oversimplified by ignoring them. Thank you for patiently pointing them out.

**Comment by snowystinger** [comment]
Hey there @agriffis , we've started our conversation with Design. Could you give some more specifics of the kinds of scales you need? We would like to gather some data to help explore the use cases.

**Comment by agriffis** [comment]
Here are two examples:

1. "How many employees in your organization?"

For this question, you want to provide high precision for answers at the low end, because it *matters* to the small biz that they can answer exactly. For large biz, they don't care about precision, they just need to be in the ballpark.

Psychologically, it's affirming for the small biz that their answer of "8" isn't buried at the far left of their slider, even though the slider can accommodate 100,000.

For this kind of slider, you might want log2 or log10 depending on how you want the curve to work.

2. "How many calories do you typically eat in a day?"

In this example, you want a bell curve. Answers under 1,000 or over 4,000 are atypical but you need to accommodate them. Most of the answers are near the middle and you'd like to have the most precision and usability in that portion of the slider.

**Comment by snowystinger** [comment]
Thanks so much for the extra information. Design has made a decision. In Spectrum components we'll only be supporting linear and log scales. In Aria/Stately though, in order to implement and future proof this, we'll support all scales.
If you're interested in contributing that for the hooks so you can use it, that'd be awesome :)
We'll start with the scale props `scale?: {transform: (val:number) => number, inverse: (val:number) => number}` for the hooks and see how we like it there.
Expectation being that value/min/max/step should all be the actual value
inverse function to map the actual value to a percentage (linear)
transform function to map from linear to the original scale for onChange

open to suggestions on names for the props, we just matched it with the terms in math for the moment
the intent being that we could then easily create some default scales and use TS to enforce that if one is provided, both must be provided.

```
import {logScale} from 'somewhere';
<Slider scale={logScale} />
```

**Comment by snowystinger** [comment]
No worries, thanks for letting us know!
Good news is, if anyone needs it, the hard work has been done. Just a matter of implementing it now.

**Comment by kazlauskis** [comment]
We also have such a requirement. Including the logarithmic progression scale option in the React Spectrum Slider would be really helpful for mobile components. Thanks!

**Comment by reidbarber** [comment]
Closing since we haven't had this use case come up yet in Spectrum, but we're open to a contribution.
